### DISCUSSION TITLE
SKRuntimeEffect shader syntax or doc?

### DISCUSSION THREAD

**Discussion by dbriard** [mono/SkiaSharp#2516] [Q&A]
Hi,
I am trying to create shaders using SKRuntimeEffect but I got errors using samples provided by Skia documentation.

1 ) For example, with the following code, I go the error below:
```
string sksl =
                    @"uniform shader image;
                    half4 main(float2 coord) {
                      return image.eval(coord).bgra;  // Sample 'image', then swap red and blue
                    }";
```
**"error: 3: type 'fragmentProcessor' does not have a field named 'eval'\n1 error\n"**

2) Another example, I would like to create a ColorFilter from a shader dans found and example here https://github.com/google/skia/blob/main/gm/runtimeshader.cpp#L264-L425 so I tried:

```
var src2 = @"
                        uniform shader color_cube;

                        half4 main(half4 inColor) {
                            float4 c = unpremul(inColor);
                            // Premul again
                            color.rgb *= color.a;
                            return color;
                        }";

                    string errorText;
                    // Parse the SkSL, and create an SkRuntimeEffect object:
                    using var effect = SKRuntimeEffect.Create(src2, out errorText);
```
and got the error: **"error: 4: 'main' parameters must be: (), (vec2), or (float2)\n1 error\n"**
Is there a sample how to create a ColorFilter from shader using Skia*Sharp*?

3) I found a working example using "in fragmentProcessor color_map;" but I have no idea where this syntax come from??? 

Any help to understand is welcome, thank you.

**Comment by mattleibow** [comment]
The syntax used in the 2.x release was quite old and not yet standardized. In the new 3.x series, they have formalized the language a bit.

You can find more:

 - https://shaders.skia.org/
 - https://skia.org/docs/user/sksl/
